# Taller 1: Introducción a Python 3
***Elaborado por Ariel Camacho***

1. Comentarios
2. Impresión y ayuda
3. Operaciones aritméticas básicas
4. Listas, tuplas y `range`
5. Operaciones con listas
6. Enunciado `if` y condicionales
7. Ciclos `for` y `while`

---


## Sobre Google Colab
Iremos explorando los fundamentos de Python usando **Google Colaboratory** ("Colab"). Para mayor información de Colab, visite el siguiente enlace: https://research.google.com/colaboratory/intl/es/faq.html

Colab es un servicio de Google que nos permite escribir y ejecutar código de Python usando recursos informáticos de Google. En la versión gratuita hay una serie de limitantes, visite el siguiente enlace para conocerlos a detalle: https://research.google.com/colaboratory/intl/es/faq.html#disallowed-activities

# Introducción a Python 3

Python 3 es un lenguaje de programación de alto nivel. Visualmente, se caracteriza por tener una sintaxis limpia y utilizar tabulación para indicar niveles de anidamiento de código. Tim Peters escribió el "Zen de Python" (https://peps.python.org/pep-0020/) que aglomera una filosofía general sobre desarrollo de código en este lenguaje de programación, dentro de lo que destaca lo siguiente:
*   Explicit is better than implicit.
*   Simple is better than complex.

Se recomienda consultar el siguiente tutorial oficial de Python 3: https://docs.python.org/es/3/tutorial/

## Comentarios en código

Es una buena práctica colocar **comentarios** en partes relevantes de nuestro código para que otras personas o nosotros mismos podamos entenderlo mejor cuando lo veamos en el futuro. Los comentarios no se ejecutarán.

Para poner comentarios en línea escribimos `#` seguido a la derecha de un comentario ilustrativo.

Para poner un bloque de comentarios escribimos `'''`, luego el comentario y cerramos nuevamente con `'''`.

In [ ]:
# Este es un comentario en línea

''' Comentario '''

' Comentario '

## Tipos elementales de datos

Los datos en Python tiene asociado un **tipo**. De los más comunes son:
* `str()` corresponde a cadenas, también se puede escribir usando comillas como `'hola'`.
* `int()` corresponde a números enteros como 4.
* `float()` corresponde a números con punto flotante como 4.13.
* `complex()` corresponde a números complejos usando el formato `a+bj` donde a y b son flotantes.

Para conocer el tipo de una variable podemos usar el comando `type()`.

In [ ]:
a = 'Hola mundo'
type(a)

str

In [ ]:
b = 4
type(b)

int

In [ ]:
c = 4.14
type(c)

float

Es posible realizar algunas conversiones o "**casting**" entre tipos de datos. Por ejemplo, `str(4)` convertirá 4 en la cadena `'4'`.

In [ ]:
str(int(4))

'4'

In [ ]:
type(4+1j)

complex

## Impresión en Python 3

Para imprimir usamos la función `print()` poniendo dentro la cadena o la variable cuyo valor queremos mostrar en pantalla.

In [ ]:
# imprimir cadena
print("hola Python")

hola Python


In [ ]:
x = 34

# imprimir el valor asignado a x
print(x)

34


Hay opciones para hacer impresiones con cierto formato específico. Una opción es mediante `f` seguido de la cadena junto con el uso de `{expresiones}`. Por ejemplo, si se quisieran mostrar números flotantes usando solamente dos decimales usaríamos `{:.2f}`.

In [ ]:
a = 1900.2345
b = 3.1415
c = -45.543
print(f'Flotantes a dos decimales: {a:.2f}, {b:.2f}, {c:.2f}')

Flotantes a dos decimales: 1900.23, 3.14, -45.54


Para mayor información sobre el uso de formatos se pueden consultar las siguientes fuentes:
* https://docs.python.org/es/3/tutorial/inputoutput.html
* https://docs.python.org/es/3/library/string.html#formatstrings

### *Ejercicio.* Practicar impresión con formato

Declare la variable `pi = 3.14159`. Después, use `print` con formato para imprimir `El número pi redondeado a dos decimales es 3.14` usando `pi`, es decir, no escriba directamente `3.14` en la cadena.

In [ ]:
# descomente y complete

# pi = ...
# print(f"...")

## Ayuda en Python

Siempre es recomendable revisar la documentación oficial de cualquier función o librería que se utilice. En ocasiones, nos será posible obtener una descripción de determinada función usando el comando `help`.

In [ ]:
help(print)

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



En este caso, podemos observar que los argumentos de `print` incluyen el valor a imprimir `value`, incluye un separador entre valores dado por `sep`, y por default se incluirá un salto de línea dado por `end='\n'`. De este modo, vemos que si en lugar de generar un salto de línea podríamos terminar con un punto y un espacio, por ejemplo.

In [ ]:
# usando print con otra terminación
print("Esta es una oración", end='. ')
print("Esta es una segunda oración.")

Esta es una oración. Esta es una segunda oración.


## Introducción a cadenas

En ocasiones deseamos imprimir cadenas que contengan cierto texto específico y que en cierta porción incluyamos el valor de cierta variable. Podemos **concatenar** dos cadenas utilizando el símbolo `+`.

In [ ]:
cadena1 = "hola"
cadena2 = "mundo"
cadena3 = cadena1 + cadena2
print(cadena3)

holamundo


Si tenemos un número, podríamos hacer casting para concatenar en una cadena dada.

In [ ]:
edad = 50
print('Mi hermano tiene ' + str(edad) + ' monedas.')

Mi hermano tiene 50 monedas.


In [ ]:
print(f'Mi hermano tiene {edad} monedas.')

Mi hermano tiene 50 monedas.


Dada una cadena, podemos usar el comando `len` para conocer la longitud de la misma.

In [ ]:
cadena1 = 'algoritmo'

print(f"La palabra {cadena1} tiene {len(cadena1)} caracteres.")

La palabra algoritmo tiene 9 caracteres.


Las cadenas en Python cuentan con varios métodos útiles para diversas situaciones, como unir toda una lista de cadenas o poner todas las letras en mayúsculas. Pueden consultar la documentación en el siguiente enlace para conocer esas interesantes funcionalidades:
* https://docs.python.org/3/library/stdtypes.html#string-methods

---

# Operaciones aritméticas básicas en Python 3

Las operaciones aritméticas entre números se realizan con los siguientes símbolos en Python 3:

- Suma y resta: `+, -`
- Multiplicación y división: `*, /`
- Exponenciación: `**` (cuidado entre `*` y `**`)
- Valor absoluto: `abs()`


**Ejemplo.** Imaginemos que queremos calcular $|3a-2b^4 + 0.5|/40$ donde $a=2$ y $b=3$.

In [ ]:
a = 2
b = 3
resultado = abs(3*a - 2*b**4 + 0.5) / 40

print( resultado )

3.8875


Como comentario: la suma de dos `int` en Python da como resultado un `int`. Al sumar un `int` con un `float`, el resultado es un `float`. Esto es de destacarse pues en Python 2 la convención era que `int`+`float` daba `int`.

In [ ]:
print( type(2+3) )
print( type(2+3.0) )

<class 'int'>
<class 'float'>


## Paquetería `math`

Para realizar otras operaciones y constantes especiales necesitaremos otros módulos como `math`. Para importar el módulo debemos ejecutar primero `import math`, y para usar sus atributos y métodos necesitaremos escribir `math.` seguido del atributo o método.

In [ ]:
# importando el módulo "math" para usar sus funciones
import math

Algunas de las funciones y constantes incluídas en `math` son las siguientes:

|método|descripción|
|---|---|
|math.sqrt()|raíz cuadrada|
|math.exp()|función exponencial|
|math.log()|logaritmo natural|
|math.log10()|logaritmo base 10|
|math.sin()|función seno|
|math.cos()|función coseno|
|math.asin()|función arcoseno|
|math.acos()|función arcocoseno|
|math.tan()|función tangente|
|math.floor()|función piso|
|math.ceil()|función techo|
|math.factorial()|factorial|

|atributo|descripción|
|---|---|
|math.pi|número $\pi$|
|math.e|número $e$|

In [ ]:
help(math.sqrt)

Help on built-in function sqrt in module math:

sqrt(x, /)
    Return the square root of x.



Para mayores detalles de la librería `math`, se puede revisar la documentación: https://docs.python.org/3/library/math.html

In [ ]:
# raíz cuadrada de x=2
x=2
print(math.sqrt(x))

1.4142135623730951


In [ ]:
help(math.cos)

Help on built-in function cos in module math:

cos(x, /)
    Return the cosine of x (measured in radians).



In [ ]:
print( math.cos(math.pi/3) )

0.5000000000000001


### *Ejercicio:* Determinar el ángulo

1. Use `help` para conocer un poco sobre la función `acos` del módulo `math`.
2. Use la función `acos` del módulo `math` para encontrar el ángulo $\alpha$ en grados tal que $\cos(\alpha)=1/2$.

In [ ]:
# help( ... )
# ...

# salida esperada: algo cercano a 60

---

# Listas y tuplas

En Python 3 tenemos dos estructuras de datos básicas:
1. **Listas:** son "mutables", es decir, es posible modificarlos una vez que han sido creados. Se usan paréntesis cuadrados `[ ]` para declararlos, poniendo dentro sus elementos.
2. **Tuplas:** son "inmutables", no se pueden modificar una vez creados. Se usan paréntesis normales `( )`.

In [ ]:
lista1 = [3,2,1,4] # lista
tupla1 = (5,2,3,2) # tupla

In [ ]:
print( type(lista1) )
print( type(tupla1) )

<class 'list'>
<class 'tuple'>


Así como con las cadenas, es posible usar `len` para saber cuántos elementos tiene una lista o una tupla.

In [ ]:
print(len(lista1), len(tupla1))

4 4


## Índices

Para acceder a un elemento de una lista o tupla usamos `variable[índice]`. En Python, **el primer índice es 0**.

In [ ]:
print( lista1[0] ) # primer elemento de lista1
print( tupla1[3] ) # cuarto elemento de tupla1

3
2


Podemos usar índices negativos para "ir de derecha a izquierda". Así, el índice `-1` nos da el último elemento de la lista o tupla.

In [ ]:
print( lista1[-1] ) # último elemento de lista1
print( tupla1[-2] ) # penúltimo elemento de tupla1

4
3


## Modificando listas

Es posible modificar (reasignar valores) a elementos específicos de listas. Para ello hacemos una declaración usual como lo haríamos con otras variables.

In [ ]:
lista2 = [3,6,7,8,3]
lista2[2] = -4 # asignar -4 a a7[2]
print(lista2)

[3, 6, -4, 8, 3]


Al intentar hacer esto con tuplas, obtendremos un error pues las tuplas son inmutables.

In [ ]:
# a2[0] = 0

```
---------------------------------------------------------------------------

TypeError                                 Traceback (most recent call last)

<ipython-input-16-402f04ae8eda> in <cell line: 1>()
----> 1 a2[0] = 0

TypeError: 'tuple' object does not support item assignment
```

### Slicing

Podemos usar "**slicing**" para tomar partes de una lista o tupla. El formato es `variable[a:b]` donde a y b son índices permitidos con `a<b`. El resultado tomará desde el índice a hasta el índice b-1. ***Observe que no tomará hasta el índice b sino hasta uno antes***.


In [ ]:
# Objetivo: imprimir primeros dos elementos de lista1
print( lista1[0:2] )

[3, 2]


In [ ]:
# objetivo: imprimir elementos 2do,3ro, y 4to de tupla1
print( tupla1[1:4] )

(2, 3, 2)


Hay un tipo especial en Python 3 llamado **range**. Pueden pensarlo como una lista aunque en realidad no lo sea. Es útil para realizar ciclos `for`.

Su formato es `range(n)` donde n es un entero positivo, y lo podemos imaginar como "la lista $(0,1,...,n-1)$" aunque en realidad no lo sea.

In [ ]:
miRango = range(5) # una "lista" desde 0 hasta 4
type(miRango)

range

## Agregando y quitando elementos a una lista

Para agregar elementos al final de una lista, se puede usar `+` ó `append`.

In [ ]:
lista3 = [1]
lista3.append(4) # agregar 4 al final de a9
print(lista3)

[1, 4]


In [ ]:
lista3 = [1]
lista3 = lista3 + [4] # agregar 4 al final de a8, y reasignar a a8
print(a8)

[3, 4]


Para las tuplas no es posible agregar o quitar elementos. Se debería hacer una tupla diferente.

Para quitar elementos de una lista podemos usar `remove` o `pop`.

In [ ]:
help(list.remove)

Help on method_descriptor:

remove(self, value, /)
    Remove first occurrence of value.
    
    Raises ValueError if the value is not present.



In [ ]:
help(list.pop)

Help on method_descriptor:

pop(self, index=-1, /)
    Remove and return item at index (default last).
    
    Raises IndexError if list is empty or index is out of range.



In [ ]:
lstNombres = ["Ana", "Beto", "Carla", "Diego"]
lstNombres.remove("Beto") # eliminará a "Beto" de la lista
print( lstNombres )

['Ana', 'Carla', 'Diego']


In [ ]:
lstNombres = ["Ana", "Beto", "Carla", "Diego"]
lstNombres.pop(2) # eliminará a "Carla" de la lista
print( lstNombres )

['Ana', 'Beto', 'Diego']


Hay mucho más que explorar sobre listas, como **ordenar**, **insertar** e incluso la gran herramienta de **comprensión de listas**. Se invita a continuar revisando fuentes como la documentación oficial: https://docs.python.org/3/tutorial/datastructures.html

### *Ejercicio.* Practicando con listas

1. Haga una lista `lstFibo` que tenga a los elementos `1, 1, 2, 3`.
2. Use `append` ó `+` junto con el uso de índices negativos para agregar la suma de los últimos dos elementos de la lista.
3. Repite el paso anterior.
4. Imprima el total de elementos que tiene `lstFibo`.
5. Use slicing para imprimir los últimos tres elementos de la lista.

---

# Condicionales

Hay un tipo de variable para **booleanos** dado por `bool()`. Puede tener dos posibles valores: `True` o `False`, que tienen como equivalencia los valores respectivos de 1 ó 0.

## Conectores lógicos

Los conectores lógicos básicos en Python son `not`, `and` y `or`, los cuales siguen las mismas reglas que en lógica. Por ejemplo, `True and False` es `False`, mientras que `True or False` es `True`.

In [ ]:
not(True and (False or False))

True

## Comparaciones

Es habitual necesitar identificar casos para ejecutar código específico para cada uno de ellos. En particular, estos casos suelen ser de los siguientes tipos:
1. **Igualdad:** `a == b`
2. **No igualdad:** `a != b`
3. **Mayor o menor:** `a < b`, `a > b`, `a <= b`, `a >= b`
3. **Pertenece:** `a in LISTA`

In [ ]:
miLista = [0, 1, 2, 0]

bool1 = miLista[0] == miLista[1]
bool2 = miLista[0] == miLista[-1]
bool3 = 3 in miLista
bool4 = 2.0 in miLista

print(bool1, bool2, bool3, bool4)

False True False True


## Enunciados "if"

El esquema básico para utilizar `if` en Python es el siguiente:


```
if (condicional):
  código         # <- observe indentación
```

**Nota:** En Python 3 se usan **indentaciones** (como cuando se presiona la tecla Tab) para anidar código subordinado. Así, justo debajo de `if ... :` la siguiente línea debe iniciar con una indentación presionando la tecla Tab y luego ya se escribe el código a ejecutarse dentro del enunciado `if`. En caso de querer más de una línea de código dentro del enunciado `if` lo importante es **mantener el mismo nivel de indentanción**. Si se reduce el nivel de indentación saldremos del enunciado `if`.

In [ ]:
# en este caso entrará al "if"
palabra1 = "algoritmo"

if len(palabra1) > 8:
  print(f"la palabra \'{palabra1}\' es muy larga")

# en este segundo caso no entrará al "if"
palabra2 = "derivada"

if len(palabra2) > 8:
  print(f"la palabra \'{palabra2}\' es muy larga")

la palabra 'algoritmo' es muy larga


# Ciclos `for` y `while`

Los ciclos básicos son `for` y `while`. Para `for` uno de sus formatos básicos es:
~~~
for (variable) in (range, lista o tupla):
  código
~~~



In [ ]:
# objetivo: imprimir los números del 0 al 10

for contador in range(11):
  print(f"{contador}, ", end='')

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 

Para `while` tenemos la plantilla:
~~~
while (condicion):
  código
~~~

Hay que procurar actualizar la variable de la condición de forma que no se genere un bucle infinito.

In [ ]:
iter = 1
while iter < 10:
  print(f"paso {iter} ", end='')
  iter = iter + 1

paso 1 paso 2 paso 3 paso 4 paso 5 paso 6 paso 7 paso 8 paso 9 

### *Ejercicio.* Suma de pares, impares y cuadrados

Trata de utilizar `print`, `for` o `while` para imprimir las siguientes sumas:
1. De los primeros 20 enteros positivos.
2. De los enteros positivos pares menores o iguales a 20.
3. De los enteros positivo impares menores que 20.

Puede ponerlos en distintos bloques de código.

**Nota:** En caso de usar `while` asegure que está actualizando la variable de la condición y que no se vaya a generar un bucle infinito.

> **Salidas esperadas:** 210, 110 y 100.



### *Ejercicio.* Fibonacci nuevamente

1. Declare `lstFiboFor = [1, 1]` y `lstFiboWhile = [1, 1]`.
2. Por un lado, haga un ciclo `for` para agregar 20 términos de la sucesión de Fibonacci a `lstFiboFor`.
3. Por otro lado, haga un ciclo `while` para agregar los términos de la sucesión de Fibonacci a `lstFiboWhile` que sean menores a 3000.
4. Imprima el último elemento de `lstFiboFor` y de `lstFiboWhile`.

In [ ]:
# lstFiboFor = ...
# lstFiboWhile = ...

# for n in ... :



# while ... :




## Enunciados `else`, `break` y `continue`



Al usar `if` podemos complementar con un `else` al mismo nivel para ejecutar código en caso de que la condición en el `if` correspondiente sea `False`.

In [ ]:
sum1=0
sum2=0

for n in range(1,11):
  if n<5:
    sum1=sum1+n
  else:
    sum2=sum2+n
  print(f"sumando {n} ", end='')

print("\n", sum1, sum2)

sumando 1 sumando 2 sumando 3 sumando 4 sumando 5 sumando 6 sumando 7 sumando 8 sumando 9 sumando 10 
 10 45


Por otro lado, en ocasiones necesitaremos terminar anticipadamente un ciclo. Para hacerlo podemos usar `if` junto con `break`.

In [ ]:
for n in list("algoritmo"):
  if n == "r":
    print("\nse encontró una \'r\', saliendo")
    break
  print(n+"-",end='')

a-l-g-o-
se encontró una 'r', saliendo


Y a veces necesitaremos saltar algunos casos en un ciclo continuando con los siguientes. Para esta podemos emplear `continue`.

In [ ]:
for n in list("algoritmo"):
  if n == "o":
    continue  # si encontramos una 'o', romper ejecución actual y pasar a siguiente iteración
  print(n+"-",end='')

a-l-g-r-i-t-m-

### *Ejercicio.* Aplausos y vueltas

1. Declare `aplausos = 0`
2. Haga un ciclo `for` que ejecute código para un contador `n` que recorre desde 0 hasta 99.
3. Dentro del ciclo, imprima el valor de `n` si 2 ó 3 no son dígitos del valor de `n`.
4. Si 3 es dígito del valor de `n`, aumente el valor de `aplausos` en 1 y y pase al siguiente valor de `n`.
5. Si 2 es dígito del valor de `n`, imprima `shh` y pase al siguiente valor de `n`.
6. Si el valor de aplausos llega a 15, rompa el ciclo `for`.

In [ ]:
# @title
aplausos = 0
for n in range(100):
  if '3' in str(n):
    aplausos += 1
    if aplausos == 15:
      break
    continue
  if '2' in str(n):
    print('shh',end=' ')
    continue
  print(n,end=' ')

0 1 shh 4 5 6 7 8 9 10 11 shh 14 15 16 17 18 19 shh shh shh shh shh shh shh shh shh 40 41 shh 44 45 46 47 48 49 50 51 shh 